File: sample_py/scope.py
'''
Author: avert-win
Date: 2022-04-29 20:52:19
LastEditTime: 2022-07-09 21:54:55
FilePath: \ProgramDaily\sample_py\scope.py
Description: 简介
'''

#% LEGB: local, enclosed, global, built-in. 依次进行寻找，前者可覆盖后者

#% parenthese, square brackets(brackets), curly brackets(curly), angle brackets

#! 1. 函数内赋值不改变参数，但列表、集合等的(局部)修改可以 
#? 哪些数据类型可以在函数内部使用哪些方式改变？
#Q: 实际上是引用的问题，只要是数据的局部，并且是非复制方式，就可以
#   在函数内改变；而整个变量名直接赋值则不可（因为引用计数，只是改变了引用）。
from attr import attr
from numpy import squeeze

def func1(arg1):
    print(f"\nget param arg1: {arg1}")
    arg1 = [1,2,3]
    print(f'change arg1 to: {arg1}')

arg1 = 'global arg1.'
func1(arg1)
print(f'exit function, arg1 is:   {arg1}')


#! 2. global关键字
#% This is wrong because declaring global variable is an entire statement.
# global arg2 = 'arg2 is also a string'

#% global 关键字是在函数内部使用的，声明函数内部的该变量来自全局变量，在此处声明没有用。
global arg2
arg2 = 'arg2 in global'

def func2():
    #%如此声明使用的变量来自全局scope.
    global arg2
    # arg2 = 'arg2 in local'
    print(f'In func2, arg2:   {arg2}')
    arg2 = [1,2,3]
    print(f'change arg2 in function:   {arg2}')

func2()
print(f'global arg2 is: {arg2}')


#! 3. nonlocal关键字和闭包函数
# 在嵌套函数中定义的，既非本地，又非全局或内建，因此称为nonlocal.
# 注意python的解释器搜索顺序是本地、非本地、全局、内建。
# 和global一样，
# nonlocal关键字的作用是标识此地符号非本地符号，因此可以重新定义或者修改而不影响本地变量。

def func3(exp):
    def power(base):
        return base**exp
    return power   # without parentheses, return a function(address and type)

# declare an (enclosed) function by using an outer function.
square = func3(2)
cube = func3(3)
print(square(10))
print(cube(10))

#% LEGB之外：Comprehensions, Exception blocks, Classes and instances.

#! 1. Comprehension Variables Scope, 推导式范围，通常用中括号或大括号包围起来

[item for item in range(5)]
# item is not accessible after list comprehension.
# print(item)

for item in range(5):
    None
# but you can get the last item(value) after for loop.
print(item)


#! 2. Exception Variables Scope, you can assign global variables there, but
#! variables there cannot being accessed outside.

list1 = [1,2,3]

ex = None
try:
    list1[3]
except IndexError as error:
    ex = error
    print(error)

# NameError: name 'error' is not defined.
# print(error)

# OK: you can assign a global variable in exception scope.
print(ex)

#! 3. Class and Instance Attributs Scope

#% 1. class不为method创建enclosed scope，在类内必须通过self传参来修改类的属性、调用类的方法；
#% 2. class具有共同的一些属性，同时每个实例还可以为自己增加属性，包括数据和方法。

class cls:
    attr = 100
    def cls_func(self):
        self.attr = 200
        return self.attr
        # 不加self直接访问的是function attrib而非类的属性
        # return attr

# 不用实例化也能直接访问静态属性
print(cls.attr)

# 函数需要参数self，因此不实例化不能访问
# print(cls.cls_func())

# 这种方法是命了别名而不是初始化！
# obj = cls
obj = cls()
print(obj.cls_func())

# 给类的实例增加一个新(数据)属性
obj.insnewattr = 3
print(obj.insnewattr)

# 给类的实例增加一个方法，不要加括号，如果加了括号就是将返回值作为属性
cls.newattr = func1
obj.newattr()

#% 可以给类或者类的实例动态增加属性(包括数据和方法)，
#% 給实例增加的仅限于该实例使用，给类增加的往后的类都能使用。

obj2 = cls()

# 访问给其它实例增加的属性，报错
# print(obj2.insnewattr)

# 访问给类增加的属性，正常执行
obj2.newattr()

File: sample_c_cpp/link/scope.c
/*
 * @Author: avert-win
 * @Date: 2022-05-06 18:28:42
 * @LastEditTime: 2022-05-07 15:00:35
 * @FilePath: /ProgramDaily/cppSample/link/scope.c
 * @Description: 简介
 * @LastEditors: avert-win
 */

/**
 * 1. Global: 由模块m定义，能被其它模块引用的符号。
 * 2. External: 由其它模块定义并被模块m引用的全局符号。
 * 3. Local: 由模块m定义和引用的本地符号（外部不能引用），例如static变量和函数
 * 
 * 强符号：函数和已经初始化的全局变量；
 * 弱符号：未初始化的全局变量（初始化为0的整形变量）
 * 
 * 对于全局符号的多重定义，有如下规则：
 * 1. 不允许有多个强符号，若有，则直接报错；
 * 2. 允许多个相同弱符号，在相同的弱符号和强符号之间，选择强符号；
 * 3. 在多个相同的弱符号之间，任意选择一个（视编译器和其他因素而定）。
 * 
 * ? 注意：编译时的本地符号，在链接器眼里是不存在的符号，链接器只关心文件整体，
 * ? 函数内部和循环内部的符号相关问题由所属文件自行解决。
 * ? 链接器的本地符号指的是static，其它文件不可访问的符号。
 * ? 链接符号和程序语言代码符号不同，多个函数内的同名static变量在链接时是不同的。
 */

static int x = 20;

// Q: 为什么此处定义了未初始化的弱符号，而全局有已初始化的强符号，但函数返回的仍然是局部的弱符号？
int f()
{
    static int x;
    return x += 1;
}

int g()
{
    static int x=10;
    return x += 3;
}

int h()
{
    return x += 5;
}

int main()
{
    printf("f(x) = %d\n", f());
    printf("f(x) = %d\n", f());
    printf("g(x) = %d\n", g());
    printf("g(x) = %d\n", g());
    printf("h(x) = %d\n", h());
    printf("h(x) = %d\n", h());
}
